### ISSUE TITLE
Provide mode (single, cluster)-specfic hooks for conditionally evaluating config methods

### ISSUE THREAD

**Issue by midnight-wonderer** [puma/puma#3571]
**Is your feature request related to a problem? Please describe.**
I am loading JVM into Ruby's process using [RJB (Ruby-Java bridge)](https://github.com/arton/rjb).
JVM is not fork-safe, so I have to load it after fork. Hence, inside Puma's `on_worker_boot`.
I need a way to set `on_worker_boot` in `puma.rb` only if Puma is forking.

**Describe the solution you'd like**
Puma does not directly cause my problem. But it would be nice if I can
* Detect modes (cluster mode, single mode, etc.) inside `puma.rb` without relying on indirect methods (e.g., environment variables) so that I can set `on_worker_boot` in the appropriate scenario and probably `on_booted` otherwise. (Or other ways to foretell the forking from `puma.rb`.)
* To solve the issue above, I must also detect if the Rails app runs inside Puma at all (e.g., in Rails console). Is this already doable?  
This is a question somewhat unrelated to the feature request. (I have some ideas, but I also want to find some opinions or recommendations when my ideas feel like hacks.)

**Additional context**
It seems this chunk of code does something similar to what I am looking for. However, it is not part of the public API.
https://github.com/puma/puma/blob/2c9aafef0ea42058bf3bb03f53805680fffecc60/lib/puma/dsl.rb#L1433-L1434

**Comment by nateberkopec** [comment]
We [warn you](https://github.com/puma/puma/blob/94461f410c444486c4516198a1379e0999bae242/lib/puma/dsl.rb#L750) if you try to set on_worker_boot in single mode, but you can just do it. Puma won't execute that hook so there's no reason not to just set both in puma.rb.

So just set both?

**Comment by midnight-wonderer** [comment]
> So just set both?

Correct me if I misunderstood.
The `on_booted` hook will be fired on the master process in clustered mode, loading unused JVM instances into it.
There might be some other aspects, too, but I did not initially explore this option mainly because of that.

**Comment by joshuay03** [comment]
> The `on_booted` hook will be fired on the master process in clustered mode, loading unused JVM instances into it. There might be some other aspects, too, but I did not initially explore this option mainly because of that.

I don't think so. As the name suggests (although it's alias `before_worker_boot` is more descriptive), it will be run inside each worker process (i.e. after fork) before the app is loaded: https://github.com/puma/puma/blob/52757c8bf96bdc07044b7243c0851a46c06c5fec/lib/puma/cluster/worker.rb#L57

Note that if using `preload_app!` the app would have already been loaded by this point.

**Comment by midnight-wonderer** [comment]
@joshuay03 Thanks for the reply.  
Actually, I tested before my reply to Nate. Here is how you can reproduce it:

puma.rb
~~~Ruby
on_worker_boot do
  puts "on worker boot: #{::Process.pid}"
end
on_booted do
  puts "on booted: #{::Process.pid}"
end
~~~

Then run Puma in cluster mode.

> [41] Puma starting in cluster mode...
> [41] * Puma version: 6.5.0 ("Sky's Version")
> [41] * Ruby version: ruby 3.3.6 (2024-11-05 revision 75015d4c1f) +YJIT [x86_64-linux]
> [41] *  Min threads: 3
> [41] *  Max threads: 3
> [41] *  Environment: production
> [41] *   Master PID: 41
> [41] *      Workers: 3
> [41] * Preloading application
> [41] * Listening on http://0.0.0.0:3000
> [41] Use Ctrl-C to stop
> 
> ...
> 
> on worker boot: 80
> [41] - Worker 1 (PID: 80) booted in 2.24s, phase: 0
> on worker boot: 77
> [41] - Worker 0 (PID: 77) booted in 2.27s, phase: 0
> on worker boot: 85
> [41] - Worker 2 (PID: 85) booted in 2.29s, phase: 0
> on booted: 41

**Comment by joshuay03** [comment]
> @joshuay03 Thanks for the reply. Actually, I tested before my reply to Nate. Here is how you can reproduce it:

My bad, I misread, I was referring to `on_worker_boot` (`before_worker_boot`), not `on_booted`.

I see what you mean now, if you define both they'll both run in cluster mode, which is not what you want.

If you're using a config file, the easiest (and somewhat conventional) way to conditionally define cluster mode only code is:
```ruby
worker_count = Integer(ENV["PUMA_WORKERS"] || 0)

if worker_count > 1
  workers worker_count

  # cluster specific code
end
```

**Comment by midnight-wonderer** [comment]
This is exactly what I came here for. (__TL;DR__ below)  

In a Rails project, the number of workers can be controlled by `WEB_CONCURRENCY` without referencing the variable in the config file at all.  
Historically,  code similar to yours (one that references `ENV["WEB_CONCURRENCY"]`) was included in the config file, but this is no longer the case. However, the environment variable is still effective.

I am currently using some heuristics for detection, too, which works fine in my specific environment.  
At worst, I can set another environment variable in the containers.  
But if there’s a more reliable way to determine this—and since `Puma` must know this information one way or another—I would prefer to use the same method.

I started this discussion here, after learning that we have `on_booted` in [a similar fashion](https://github.com/puma/puma/issues/1230).  
Introducing new APIs might take time to discuss, and not many people are loading a JVM into Ruby. Still, it's better to start the conversation now than to postpone it.

### TL;DR
Thank you for helping me find a solution, we have similar ideas. But would it be better if...

**Comment by joshuay03** [comment]
Fair enough, but imo there's no easy way to expose this via the DSL seeing as puma can only really determine if it should run in cluster mode after all configs have been loaded. Take the code you referenced in your description for example: https://github.com/puma/puma/blob/2c9aafef0ea42058bf3bb03f53805680fffecc60/lib/puma/dsl.rb#L1433-L1434

This runs as soon as a hook is defined. Let's assume the only way you're configuring puma is via the config file i.e. no `WEB_CONCURRENCY` etc. If you call `workers` after you defined the hook, the warning won't be logged:

```ruby
on_worker_boot do
  # ...
end # warning is incorrectly logged

# here the DSL doesn't know puma will eventually run in cluster mode

workers 5 

# here it does
```

Maybe instead of a DSL method that simply returns whether or not puma will run in cluster mode, it might be possible to introduce two more hooks, something like:

```ruby
single do
  # ...
end

cluster do
  # ...
end
```

which can be executed after all configs have been evaluated, by which time puma will know for sure which mode it will run in.

If this is accepted, it might also be an acceptable breaking change for v7 to enforce that cluster mode only methods are only defined inside the `cluster` hook. AFAIK there are no single mode only methods, but the same logic applies. And we would also want to enforce that `workers` can't be called inside these. Thoughts? @nateberkopec

**Comment by nateberkopec** [comment]
The single/cluster DSL proposal isn't bad... a lot of people run Puma in single mode locally but cluster in prod (I think they should be using cluster in both but whatever, certainly they run in single mode in tests for instance).

The problem is that I'm not sure who would really use it, and for what, outside of what OP has already admitted is a bit of a niche use case?

**Comment by joshuay03** [comment]
> a lot of people run Puma in single mode locally but cluster in prod (I think they should be using cluster in both but whatever, certainly they run in single mode in tests for instance).

I agree! But ime single mode tends to get picked to make debugging easier. Having a high worker timeout in dev cluster helps with that, but doing things like attaching a remote debugger to the correct process for example is still tricky. I don't see this changing anytime soon seeing as Rails has settled on this approach as well.

> The problem is that I'm not sure who would really use it, and for what, outside of what OP has already admitted is a bit of a niche use case?

At two of my previous workplaces we did something like:
```ruby
worker_count = Integer(ENV["PUMA_WORKERS"] || 0)
cluster_mode = worker_count > 1

if cluster_mode
  workers worker_count

  # cluster specific code
end
```

so that
1. the config is easier to grok (non-puma savvy folks don't need to read through the dsl docs to know which of the configs are cluster specific)
2. the [defined-hook-in-single-mode warnings](https://github.com/puma/puma/blob/801d292e174d1b32b7b95bafef498ea44c8f2a83/lib/puma/dsl.rb#L1430-L1443) aren't logged in dev

I think it would be nice if users didn't have to figure that out themselves:
```ruby
workers Integer(ENV["PUMA_WORKERS"] || 0)

cluster do
  # cluster specific code
end
```

Difference here is `#workers` is called even if `PUMA_WORKERS` is 1, which is advised against. But I think it's fair to say that the only scenario where this could unintentionally happen is if automatically setting `PUMA_WORKERS` based on vCPU count, or similarly doing something like `worker_count = Integer(ENV["PUMA_WORKERS"] || Concurrent.processor_count)`, which is unlikely to be 1 nowadays. This is the same reasoning behind `WEB_CONCURRENCY=auto`.

Some examples from the wild: https://github.com/search?q=language%3ARuby+path%3A%2Fpuma.rb%2F+%2F%5Cnif.*worker.*%3E.*1%2F+&type=code

**Comment by nateberkopec** [comment]
tbh for me your original approach of "workers = 1 actually means single mode" is confusing, so the drawback of 

>Difference here is #workers is called even if PUMA_WORKERS is 1,

... isn't really a drawback.

I think I'm OK with a `cluster/single` DSL addition. Anyone else have comments or ideas?

**Comment by dentarg** [comment]
> a `cluster/single` DSL addition

What are we talking about? Maybe it should be outlined in a new issue

**Comment by nateberkopec** [comment]
I think it solves the original problem of OP so we can keep it in this issue.

```ruby
cluster do 
  # only runs if worker >= 1
end 

single do 
  # only runs if worker < 1
end
```

**Comment by midnight-wonderer** [comment]
> I think I'm OK with a `cluster/single` DSL addition. Anyone else

This is how I understand the DSL based on the context of this discussion:
~~~Ruby
single do
  # Will run in single mode
end

cluster do
  # Will run on each worker in cluster mode
end

# Note: The comments above
~~~

I could be mistaken, but the `cluster` functionality may need further clarification. The `single` block is quite straightforward, but for my use case, this alternative also works:

~~~Ruby
single do
  ::Rjb::load(classpath, java_options)
end

on_worker_boot do
  ::Rjb::load(classpath, java_options)
end
~~~

Is the cluster block essentially equivalent to `on_worker_boot`, but with the warning suppressed?
If it's similar, should we simply add an option to `on_worker_boot` to suppress the warning?

Or is it something more akin to `on_booted`?

**Comment by nateberkopec** [comment]
I was thinking more "the configuration described within this block will only apply if workers is gteq/less than 1", not that the code within the block would run at a particular time, i.e. like on_worker_boot.

**Comment by midnight-wonderer** [comment]
Ah! I see.

So in practice, it can be nested
Like this:
~~~Ruby
cluster do
  on_worker_boot do
    ::Rjb::load(classpath, java_options)
  end
end
~~~

That works even better.

**Comment by dentarg** [comment]
What's a good title for this issue?

I don't think the current one correctly describes the task at hand

**Comment by joshuay03** [comment]
> What's a good title for this issue?

If we're going off of the original request:

'Provide an API to detect mode (single, cluster) in which config is evaluated'

If we're going off of the proposed solution:

'Provide mode (single, cluster)-specfic hooks for conditionally evaluating config methods*'

*doesn't have to be config methods, could really be anything, but that's the intended use here

**Comment by joshuay03** [comment]
A note for whoever picks this up. I don't know if this will be targeted for v7 or before, but it's probably a good idea to base your work on top of https://github.com/puma/puma/pull/3297. These hooks should be evaluated just before those conditional defaults are set. I think it would be useful to extend those tests as well to verify that for example:

```
workers 2

cluster do
  prune_bundler true
end
```

doesn't enable preloading by default.

**Comment by midnight-wonderer** [comment]
Hello,
The draft is in:
https://github.com/midnight-wonderer/puma-server/commit/ab6849390163373483dda5a308e0ffb98c818004

> It's probably a good idea to base your work on top of #3297. These hooks should be evaluated just before those conditional defaults are set.

I tried to understand the pull request, but here are the reasons I did not place the hook there:
* The block may not be idempotent.
From what I understand, `set_conditional_default_options` can be called multiple times. If I run the hooks in the same place, it might lead to unintended behavior.
* The availability of the `clustered?` method.
By placing the invocations in `launcher.rb`, we can infer the actual value Puma uses to determine the mode of operation.
* The feature we're trying to add here isn't really a "hook" but rather dynamic configuration blocks.
It won't be called as part of Puma's lifecycle. The appropriate time to call the block might be once the mode of operation is settled.
This is also why I did not use the `process_hook` method.

I know this is only a draft and I don't touch the tests yet.  
But I did a simple test with `puts` method on my Rails application and it kinda works.

Please let me know what you think.
Should I continue or abandon the branch?

**Comment by joshuay03** [comment]
> * The block may not be idempotent.
>   From what I understand, `set_conditional_default_options` can be called multiple times. If I run the hooks in the same place, it might lead to unintended behavior.

Yes unfortunately the way `Configuration`s are currently used they may be provided some options during initialization and then again during `#load` (when also loading a config file for example), which iirc is why the duplicate calls were necessary.

RE: Duplicate calls, there's a similar issue reported [here](https://github.com/puma/puma/issues/3364), but reading through those comments it seems Puma's stance is that it expects configs to be idempotent.

> * The availability of the `clustered?` method.
>   By placing the invocations in `launcher.rb`, we can infer the actual value Puma uses to determine the mode of operation.

This doesn't necessarily need to be inside the launcher. As long as it's after `Configuration#load` has done its thing, it should work just fine. I don't think the launcher is appropriate for it anyway as the config may be loaded in other places, e.g. [here](https://github.com/puma/puma/blob/2bc5bb8b4ec4767537528eed891edbd08d039580/lib/puma/control_cli.rb#L131), and we should leave that responsibility to the config itself, and not whatever is utilising it.

> * The feature we're trying to add here isn't really a "hook" but rather dynamic configuration blocks.
>   It won't be called as part of Puma's lifecycle. The appropriate time to call the block might be once the mode of operation is settled.
>   This is also why I did not use the `process_hook` method.

Yes those hooks are a separate thing, and I don't think this should piggy back off their implementation. Here's what I put together: https://github.com/puma/puma/compare/master...joshuay03:puma:add-single-and-cluster-hooks#diff-122c03b9c96f060a9cc29b5ed3446467e1909b0054b4f5cfd95ee41cfff832e7R1375

> I know this is only a draft and I don't touch the tests yet. But I did a simple test with `puts` method on my Rails application and it kinda works.
> 
> Please let me know what you think. Should I continue or abandon the branch?

Up to you, I can review when you've got a PR up. Feel free to use my code if you like.

I'll revisit whether https://github.com/puma/puma/pull/3297 can avoid the duplicate call to `#set_conditional_default_options` (i.e. can it be expected that `#load` will always be called). It's been a while since I opened that and I can't really remember if/why it was necessary (or feel free to have a look yourself). However, if it is necessary there, I think the same reasoning will apply to these hooks. If not, then all we need to do (going off of my code) is remove the duplicate call in `#initialize`.

**Comment by joshuay03** [comment]
> I'll revisit whether https://github.com/puma/puma/pull/3297 can avoid the duplicate call to #set_conditional_default_options (i.e. can it be expected that #load will always be called). It's been a while since I opened that and I can't really remember if/why it was necessary (or feel free to have a look yourself). However, if it is necessary there, I think the same reasoning will apply to these hooks. If not, then all we need to do is exclude the duplicate call in #initialize.

Okay so it looks like in practice we can expect both `#load` and `#clamp` to be called by the `Launcher`. I incorrectly assumed otherwise cause the tests were not following this pattern. I've updated that PR to eliminate the duplicate call, moved the call to `#clamp` as it feels more appropriate. Those changes are dependent on https://github.com/puma/puma/pull/3616.

I can now do the same for this feature, will update my branch and send a PR soon.

<hr>

>     * I still don't figure out the functionality of `configure` vs just `call`.

It yields the user options, file options and default options to the block. But now that I think about it, that seems more like an internal API, I don't see why those should be exposed to users. I'll update to just call the block instead 👍🏽

>     * In the context of a Rails application, `Configuration#run_mode_hooks` was called only once. I still haven’t figured out how, as I can see with my own eyes, that `@config.load` is there, and the constructor should be called somewhere.

If you defined them in your config file they probably only got picked up on `#load` i.e. the second duplicate call. It's only the user options (CLI/rack handler provided) that would get resolved twice. This should no longer be a concern now that I'll be removing the duplicate calls.
